src/NUnitFramework/framework/Constraints/UniqueItemsConstraint.cs

```diff
@@ -60,5 +69,123 @@ protected override bool Matches(IEnumerable actual)
 
             return true;
         }
+
+        private bool? TryFastAlgorithm(IEnumerable actual)
+        {
+            // If the user specified any external comparer with Using, exit
+            if (UsingExternalComparer)
+                return null;
+
+            // If IEnumerable<T> is not implemented exit,
+            // Otherwise return value is the Type of T
+            Type memberType = GetGenericTypeArgument(actual);
+            if (memberType == null || IsHandledSpeciallyByNUnit(memberType))
+                return null;
+
+            string methodName = nameof(ItemsUnique);
+            bool makeGeneric = true;
+
+            // Special handling for ignore case with strings and chars
+            if (IgnoringCase)
+            {
+                if (memberType == typeof(string))
+                {
+                    methodName = nameof(StringsUniqueIgnoringCase);
+                    makeGeneric = false;
+                }
+                else
+                if (memberType == typeof(char))
+                {
+                    methodName = nameof(CharsUniqueIgnoringCase);
+                    makeGeneric = false;
+                }
+            }
+
+            MethodInfo method = GetType().GetMethod(methodName, BindingFlags.Static | BindingFlags.NonPublic);
+            if (method == null)
+                throw new InvalidOperationException("Internal error - Method not found: " + methodName);
+            
+            if (makeGeneric)
+                method = method.MakeGenericMethod(memberType);
+
+            return (bool)method.Invoke(null, new object[] { actual });
+        }
+
+        private static bool ItemsUnique<T>(IEnumerable<T> actual)
+        {
+            var hash = new HashSet<T>();
+
+            foreach (T item in actual)
+            {
+                if (hash.Contains(item))
+                    return false;
+
+                hash.Add(item);
+            }
+
+            return true;
+        }
+
+        private static bool StringsUniqueIgnoringCase(IEnumerable<string> actual)
+        {
+            var hash = new HashSet<string>();
+
+            foreach (string item in actual)
+            {
+                string s = item.ToLower();
```

Comment:
We should not be using `.ToLower()` anywhere. It's expensive because it allocates a new string. The entire method can be made more general than hardcoding CurrentCultureIgnoreCase if we instead pass an `IEqualityComparer<>` to the HashSet constructor. In fact, this whole method can be removed if `ItemsUnique` starts accepting an `IEqualityComparer<T>` parameter.

Reply:
Understood, but it's what the existing `NUnitEqualityComparer` does.

However, I do like your idea, subject to my time limitations as noted.

Reply:
I agree that IEqualityComparer<> should be used for string and char to handle the case sensitivity.  It is significantly faster.  StringComparer already exists, and it is easy to write a CharComparer that does the same.  

It does not seem that it would be a breaking change.

Reply:
As I said, I like it too. I'm out of time for this issue, however.

Reply:
I may have missed this, but is using the CurrentCulture, which is the default for string.ToLower(), the correct thing to do, or should string.ToLowerInvariant() or string.ToLower (CultureInfo.InvariantCulture) be used?

Reply:
NUnitEqualityComparer currently uses CurrentCultureIgnoreCase if you set `.IgnoreCase` and (iirc) Ordinal if you don't.

Reply:
Then it appears that this is doing the same for the ignore case situation.  Sorry for the misunderstanding.

Reply:
NUnitEqualityComparer uses simple ToLower just as I did in this code. That's why I did it that way. Main objective is to produce the same results as NUnitEqualityComparer but more quickly.